### PR TITLE
Fix action server deadlock issue that caused by other mutexes locked in CancelCallback

### PR DESCRIPTION
**Pull request by KavenYau**
Resolves the deadlock issue mentioned in https://github.com/ros-planning/navigation2/issues/2273 and https://github.com/ros-planning/navigation2/pull/2304

The unit test which I added could also reproduce the issue.

### CODE REVIEW

**Comment by KavenYau** [comment]
It's also related to #1599.
```c++
        if (handle->is_active()) return rclcpp_action::CancelResponse::ACCEPT;
        else return rclcpp_action::CancelResponse::REJECT;
```
if did not add the if/else condition to CancelCallback to check whether the goal active, it would throw exception.

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
@clalancette I leave it to you if you think this bug patch should make it into Galactic, but it would be very nice if it did since it impacts a number of Nav2 users :smile: 

Love it, thanks for the patch @KavenYau! Hopefully this is the last-last race condition issue in the action server we'll need to fix :smile:

**Comment by clalancette** [comment]
So the good news is that as it stands, this patch is ABI compatible.  So we could do it in a Galactic patch release.

That said, this would be a good fix to have (and is one of the reasons to do testing at this point).  So I've put it on the Galactic project board to look at.  That doesn't guarantee we'll get it in, but we'll definitely consider it.

**Comment by SteveMacenski** [comment]
Let me know if I can help speed things up - its overall a pretty simple change so I'd hope a maintainer here could review it shortly

File: rclcpp_action/include/rclcpp_action/server.hpp
Comment (wjwwood) [inline rclcpp_action/include/rclcpp_action/server.hpp:372]: I was concerned that this introduced a race here, allowing the goal to be finished asynchronously, after it was found but before we call the user callback on this line. But it seems that was always the expectation, since the user callback can return `CancelResponse::ACCEPT` or `CancelResponse::REJECT` (though that might only be one reason for it.

This is why you need the new code in the test:

```c++
        if (!handle->is_active()) {
          return rclcpp_action::CancelResponse::REJECT;
        }
```

However, now I'm wondering about `goal_handle->_cancel_goal();`, what if this happens:

- user callback checks `if (!handle->is_active())` which fails
- user returns `CancelResponse::ACCEPT`
- (⚡) context switch, goal finishes (terminates? enters terminal state? moves to aborted/succeeded state? not sure on the terminology)
- switch back to original thread, call `goal_handle->_cancel_goal()` because `CancelResponse::ACCEPT == resp`
- now we're canceling a finished goal

It looks like `_cancel_goal` doesn't care if the state changed under us:

https://github.com/ros2/rclcpp/blob/1037822a63330495dcf0de5e8f20544375a5f116/rclcpp_action/src/server_goal_handle.cpp#L80-L88

Though `rcl` does:

https://github.com/ros2/rcl/blob/414fdbd479b83f2b8af8235ad1fa3aeba2b0027f/rcl_action/src/rcl_action/goal_handle.c#L92-L99

Maybe we should be calling `rcl_action_goal_handle_is_cancelable()` before assigning it, or catch that rcl error?

Also, it should be mentioned to users (maybe already documented) that even though their cancel goal callback is called, and they return `CancelResponse::ACCEPT`, the resulting state may not be `canceled`...

@jacobperron any insight? I'm unfortunately not as versed in the design/implementation of actions as I think I need to be to judge this.

See also: https://design.ros2.org/articles/actions.html

**Review by wjwwood — COMMENTED** [review]
I'm concerned that this will require all cancel callbacks for servers to check the `is_active()` state... which isn't ideal, imo.

Also, I'm concerned we could be allowing a goal to change states from SUCCEEDED or ABORTED to CANCELED erroneously.

I don't feel confident approving this change myself, I think we need more eyes on this.

File: rclcpp_action/include/rclcpp_action/server.hpp
Comment (KavenYau) [reply rclcpp_action/include/rclcpp_action/server.hpp:372]: @wjwwood Thanks for looking into it.

Yes, this patch could only fix the deadlock issue.

> However, now I'm wondering about goal_handle->_cancel_goal();, what if this happens:
user callback checks if (!handle->is_active()) which fails
user returns CancelResponse::ACCEPT
(⚡) context switch, goal finishes (terminates? enters terminal state? moves to aborted/succeeded state? not sure on the terminology)
switch back to original thread, call goal_handle->_cancel_goal() because CancelResponse::ACCEPT == resp
now we're canceling a finished goal

Actually this case I had mentioned in https://github.com/ros2/rclcpp/issues/1599#issuecomment-814145969, but maybe I didn't explain it clearly enough .

> Maybe we should be calling rcl_action_goal_handle_is_cancelable() before assigning it, or catch that rcl error?

> Also, it should be mentioned to users (maybe already documented) that even though their cancel goal callback is called, and they return CancelResponse::ACCEPT, the resulting state may not be canceled...

I think catching the exception internally is a straightforward way to resolve it, then the users have not to add `if (!handle->is_active())` to there CancelCallback.

But @fujitatomoya thought it was a strange behavior. https://github.com/ros2/rclcpp/issues/1599#issuecomment-810283121. So far I have no other solotions to fix this issue.

Comment (fujitatomoya) [reply rclcpp_action/include/rclcpp_action/server.hpp:372]: removing mutex before calling user callback looks okay to me 👍 

> I think catching the exception internally is a straightforward way to resolve it

i believe the point for this discussion is how it appears to be for user application. currently user application does receive exception if this happens. which i think it seems to be okay since user application gets this corner case as an exception. (probably exception message and adding check `rcl_action_goal_handle_is_cancelable` would be better.) i mean basically it does what needs to be done. 

i just do not have good feeling that system conceals this exception from user application, which is suggested on https://github.com/ros2/rclcpp/issues/1599.

Comment (KavenYau) [reply rclcpp_action/include/rclcpp_action/server.hpp:372]: I see.

> even though their cancel goal callback is called, and they return CancelResponse::ACCEPT, the resulting state may not be canceled..

In addition, I wonder make sure if you agree with this solution?

And how we fix it? catching the exception like below code? or doing some changes like this patch https://github.com/KavenYau/rclcpp/commit/3ba7b2ef01449929b6c976bdb9f21c886dff3464? or other better suggestions?
```c++
try {
  goal_handle->_cancel_goal();
} catch (const rclcpp::exceptions::RCLError & ex) {
  return CancelResponse::REJECT;
}
```

Comment (jacobperron) [reply rclcpp_action/include/rclcpp_action/server.hpp:372]: > Maybe we should be calling rcl_action_goal_handle_is_cancelable() before assigning it, or catch that rcl error?

It looks like an issue unrelated to this change. I think that it would be nice to handle this race between transitioning to canceling and some terminal state (succeeded or aborted) instead of throwing an exception.

> Also, it should be mentioned to users (maybe already documented) that even though their cancel goal callback is called, and they return CancelResponse::ACCEPT, the resulting state may not be canceled...

Indeed, by design *accepting* a cancel request only tries to transition the goal to the **canceling** state; it does not mean the goal is immediately transitioned to the **canceled** state. I don't know if this is documented in rclcpp_action very well, but I think the design doc captures it?

**Review by jacobperron — COMMENTED** [review]
The change proposed in this PR looks okay to me. Correct me if I'm wrong, but I think the race related to transitioning to canceling and transitioning to a terminal state is not a new bug being introduced (it already exists).

**Comment by KavenYau** [comment]
Is there any progress on this？ :grin:

File: rclcpp_action/include/rclcpp_action/server.hpp
Comment (fujitatomoya) [reply rclcpp_action/include/rclcpp_action/server.hpp:372]: > It looks like an issue unrelated to this change.

i think so too, this fix is unrelated.

> However, it's probably better to address this in a separate PR.

agree, issue is already opened. https://github.com/ros2/rclcpp/issues/1599

> Indeed, by design accepting a cancel request only tries to transition the goal to the canceling state; it does not mean the goal is immediately transitioned to the canceled state. I don't know if this is documented in rclcpp_action very well, but I think the design doc captures it?

i did check the action design, and i see `The list of goals in the response indicates which goals will be attempted to be canceled.` as you mentioned. besides, 



it described clearly that after accepting `cancel request` that still possible transit into `succeeded`.

so i think my assumption is wrong, it should not throw exception in this case.

**Review by fujitatomoya — APPROVED** [review]
this fix itself looks good to me.

**Comment by fujitatomoya** [comment]
@jacobperron @wjwwood it would be probably better if either of you take a look again? and about `exception problem`, i think we can discuss that on https://github.com/ros2/rclcpp/issues/1599?

File: rclcpp_action/include/rclcpp_action/server.hpp
Comment (wjwwood) [reply rclcpp_action/include/rclcpp_action/server.hpp:372]: I think we can follow up on this in https://github.com/ros2/rclcpp/issues/1599

File: rclcpp_action/test/test_server.cpp
Comment (wjwwood) [inline rclcpp_action/test/test_server.cpp:1243]: Should we have a TODO here mentioning the issue @fujitatomoya created?

```suggestion
        # TODO(KavenYau): this check may become obsolete with https://github.com/ros2/rclcpp/issues/1599
        if (!handle->is_active()) {
          return rclcpp_action::CancelResponse::REJECT;
        }
```

Or is the thinking that we'll need it always?

**Review by wjwwood — COMMENTED** [review]
changes lgtm

File: rclcpp_action/test/test_server.cpp
Comment (jacobperron) [reply rclcpp_action/test/test_server.cpp:1243]: I think we should patch things so that users don't have to check if the goal is active. IIUC, there's still potentially a race doing this in general, since the goal state may change between checking `is_active()` and returning from the user-callback.

**Review by jacobperron — APPROVED** [review]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14451/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9222/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12122/)
* Windows [Build Status](https://ci.ros2.org/job/ci_windows/14572/)

**Comment by fujitatomoya** [comment]
it seems windows job got stuck to clear out docker containers?

```
00:00:09.572 C:\J\workspace\ci_windows>powershell -Command "if ($(docker ps -q) -ne $null) { docker stop $(docker ps -q)}" 
...
```

**Comment by jacobperron** [comment]
I've retriggered the Windows build.

**Comment by fujitatomoya** [comment]
@jacobperron thanks :+1: 

all green, i think this can be merged.

**Comment by clalancette** [comment]
Should we consider backporting this one to Galactic?

**Comment by SteveMacenski** [comment]
I think so :smile:

**Comment by KavenYau** [comment]
Should / Could I backport this to Foxy / Galactic?

**Comment by clalancette** [comment]
> Should I backport this to Foxy / Galactic?

Definitely Galactic.  I haven't looked at the code in Foxy, but if it applies there and doesn't break ABI, then it can be backported to Foxy as well.

**Comment by jacobperron** [comment]
I've already added this to the Foxy project board.

**Comment by KavenYau** [comment]
> I've already added this to the Foxy project board.

Dose it mean that I don't need to make a backporting PR for it because it's already in plan?

**Comment by clalancette** [comment]
> Dose it mean that I don't need to make a backporting PR for it because it's already in plan?

It still needs someone to do the work for both Galactic and Foxy, so if you are willing to do it, it would be appreciated.

**Comment by ivanpauno** [comment]
@Mergifyio backport galactic

**Pull request merged**
